### PR TITLE
Create tools to test commands

### PR DESCRIPTION
**Pull request by LevchenkoS**
Been done:
- Added specific abstract classes for
  commands testing(CommandsTest
  and CommandRequestTest);
- Been added unit tests for base
  command classes (CommandImpl,
  CommandRequestImpl, CommandResponseImpl);
- Been added unit tests for next
  mobile response commands:
  - ListFilesResponse
  - ReadDIDResponse
  - AlertManeuverResponse
  - AlertResponse
  - SubscribeButtonResponse

Related to: [APPLINK-24911](https://adc.luxoft.com/jira/browse/APPLINK-24911)

### CODE REVIEW

**Comment by LevchenkoS** [comment]
@AByzhynar, @okozlovlux, @Kozoriz, @LuxoftAKutsan, @VVeremjova, @vlantonov, please review.

**Comment by okozlovlux** [comment]
@LevchenkoS  reviewed

File: src/components/application_manager/test/commands/command_impl_test.cc [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+    DISALLOW_COPY_AND_ASSIGN(UnwrappedCommandImpl);
+  };
+
+  // Create `SmartObject` which handle array of `SmartObjects`
+  static void CreateArray(MessageSharedPtr& msg,
+                          const size_t msg_count = kDefaultMsgCount_) {
+    if (!msg) {
+      msg = CreateMessage(smart_objects::SmartType_Array);
+    }
+    ::smart_objects::SmartArray* array = msg->asArray();
```
Comment (vlantonov) [inline src/components/application_manager/test/commands/command_impl_test.cc]: Is :: required?

Comment (LevchenkoS) [inline src/components/application_manager/test/commands/command_impl_test.cc]: @vlantonov Obviously - not. But it's more to avoid misunderstandings about "Where it comes from?".

Comment (vlantonov) [inline src/components/application_manager/test/commands/command_impl_test.cc]: In the rest of the file the smart_objects namespace is accessed without this addition. So to me it seemed that there was special reason to write this and I wondered why.

**Comment by vlantonov** [comment]
Reviewed

File: src/components/application_manager/test/commands/command_request_impl_test.cc [hunk 2]
```diff
@@ -0,0 +1,485 @@
...
+  EXPECT_CALL(app_mngr_, CheckPolicyPermissions(_, _, _, _, _))
+      .WillOnce(DoAll(GetArg3(&params), Return(kMobResultSuccess_)));
+
+  EXPECT_TRUE(command->CheckPermissions());
+  EXPECT_TRUE(params.end() !=
+              std::find(params.begin(), params.end(), kPolicyAppId_));
+}
+
+TEST_F(CommandRequestImplTest, AddDisallowedParameters_SUCCESS) {
+  const std::string kDisParam =
```
Comment (AByzhynar) [inline src/components/application_manager/test/commands/command_request_impl_test.cc]: @LevchenkoS As for me `DisParam` - is unclear. I would recommend to rename it. 
Dis... - is disabled, disallowed, dismissed or what?

File: src/components/application_manager/test/commands/command_request_impl_test.cc [hunk 3]
```diff
@@ -0,0 +1,485 @@
...
+typedef CommandRequestImplTest::UnwrappedCommandRequestImpl UCommandRequestImpl;
+typedef SharedPtr<UCommandRequestImpl> CommandPtr;
+
+TEST_F(CommandRequestImplTest, OnTimeOut_StateCompleted_UNSUCCESS) {
+  CommandPtr command = CreateCommand<UCommandRequestImpl>();
+
+  // Should be called twice:
+  // First -- on `onTimeOut` method call
+  // Second -- on destruction;
+  EXPECT_CALL(ev_disp_, remove_observer(_)).Times(2);
```
Comment (AByzhynar) [inline src/components/application_manager/test/commands/command_request_impl_test.cc]: @LevchenkoS `ev_disp_` - is event (evaluation, evacuation) display? Give more clear name to variable

File: src/components/application_manager/test/commands/command_impl_test.cc [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+  EXPECT_TRUE(command->CheckPermissions());
+  EXPECT_TRUE(command->CleanUp());
+
+  // Default value of `allowed_to_terminate_` is true
+  EXPECT_TRUE(command->AllowedToTerminate());
+  command->SetAllowedToTerminate(false);
+  EXPECT_FALSE(command->AllowedToTerminate());
+
+  const uint32_t kCorrelationId = 3u;
+  const int32_t kFunctionId = 4u;
```
Comment (AByzhynar) [inline src/components/application_manager/test/commands/command_impl_test.cc]: @LevchenkoS `int32_t = uint`?

File: src/components/application_manager/test/commands/command_impl_test.cc [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+    }
+    ::smart_objects::SmartArray* array = msg->asArray();
+    SmartObject obj;
+    for (size_t i = 0u; i < msg_count; ++i) {
+      obj[strings::app_id] = i;
+      array->push_back(obj);
+    }
+  }
+  // Create `SmartObject` which handle map of `SmartObjects`
+  static void CreateMap(MessageSharedPtr& msg,
```
Comment (AByzhynar) [inline src/components/application_manager/test/commands/command_impl_test.cc]: @LevchenkoS What if you get as parameter msg with SmartObject type INT, or String or anything else but `SmartType_Map` ? I mean it's worth to add check for incoming message smart object type before trying to add something to it.

File: src/components/application_manager/test/commands/command_impl_test.cc (hunk 1 above)
Comment (AByzhynar) [inline src/components/application_manager/test/commands/command_impl_test.cc]: @LevchenkoS What if you get as parameter msg with SmartObject type INT, bool, or String or anything else but SmartType_Array ? You will get NULL pointer here. And then in loop body you are dereferencing it. Guess what you get?

File: src/components/application_manager/test/include/application_manager/commands/command_request_test.h [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+using ::application_manager::commands::Command;
+
+template <const CommandsTestMocks kIsNice = CommandsTestMocks::kNotNice>
+class CommandRequestTest : public CommandsTest<kIsNice> {
+ public:
+  typedef typename TypeIf<kIsNice,
+                          NiceMock<MockEventDispatcher>,
+                          MockEventDispatcher>::Result MockEventDisp;
+
+  MockEventDisp ev_disp_;
```
Comment (AByzhynar) [inline src/components/application_manager/test/include/application_manager/commands/command_request_test.h]: @LevchenkoS UNCLEAR VARIABLE NAME

File: src/components/application_manager/test/include/application_manager/commands/commands_test.h [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+    return ::utils::MakeShared<Command>(msg, app_mngr_);
+  }
+
+  enum { kDefaultTimeout_ = 100u };
+
+  MockAppManager app_mngr_;
+  MockAppManagerSettings app_mngr_set_;
+
+ protected:
+  virtual void InitCommand(const uint32_t& timeout) {
```
Comment (AByzhynar) [inline src/components/application_manager/test/include/application_manager/commands/commands_test.h]: @LevchenkoS I really sorry but why do we have const ref to uint?

Comment (LevchenkoS) [inline src/components/application_manager/test/include/application_manager/commands/commands_test.h]: @AByzhynar Method `InitCommand` get invoked in template method `CreateCommand` to prepare all mocks, before the command constructor will ask for needed data. But `default_timeout` from `ApplicationManagerSettings` was return ref to its data. So to prepare it we should guarantee that the object we want to return will still be alive when we need it. But we need it outside the `InitCommand` method. So, if `timeout` will be a value, when command constructor ask for `default_timeout`, mock will return ref on dead object.

File: src/components/application_manager/test/include/application_manager/commands/command_request_test.h (hunk 6 above)
Comment (LevchenkoS) [inline src/components/application_manager/test/include/application_manager/commands/command_request_test.h]: @AByzhynar I understand that name like `event_dispatcher_` will be more clear, but, in my opinion, this name is too long.

File: src/components/application_manager/test/include/application_manager/commands/commands_test.h [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+
+enum CommandsTestMocks { kNotNice = 0, kIsNice };
+
+template <const CommandsTestMocks kIsNice = CommandsTestMocks::kNotNice>
+class CommandsTest : public ::testing::Test {
+ public:
+  typedef typename TypeIf<kIsNice,
+                          NiceMock<MockApplicationManager>,
+                          MockApplicationManager>::Result MockAppManager;
+  typedef typename TypeIf<kIsNice,
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/test/include/application_manager/commands/commands_test.h]: @LevchenkoS I think that application_manager_setting should be nice_mock in completely all cases.

File: src/components/application_manager/test/commands/command_impl_test.cc (hunk 5 above)
Comment (LevchenkoS) [inline src/components/application_manager/test/commands/command_impl_test.cc]: @AByzhynar If message have type different form `SmartType_Map` this method just simply override it to `SmartType_Map`.

File: src/components/application_manager/test/commands/command_impl_test.cc [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+        : CommandImpl(message, application_manager) {}
+
+   private:
+    DISALLOW_COPY_AND_ASSIGN(UnwrappedCommandImpl);
+  };
+
+  // Create `SmartObject` which handle array of `SmartObjects`
+  static void CreateArray(MessageSharedPtr& msg,
+                          const size_t msg_count = kDefaultMsgCount_) {
+    if (!msg) {
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/test/commands/command_impl_test.cc]: @LevchenkoS if msg is output parameter it is better to make it pointer and pass as second parameter.
Also in Unit test  it would be better to DCHECK case if you pass not empty message as output parameter.

File: src/components/application_manager/test/commands/command_request_impl_test.cc (hunk 2 above)
Comment (LevchenkoS) [inline src/components/application_manager/test/commands/command_request_impl_test.cc]: @AByzhynar If you look at test case name it will gets more clear.

File: src/components/application_manager/test/include/application_manager/commands/commands_test.h (hunk 8 above)
Comment (LevchenkoS) [inline src/components/application_manager/test/include/application_manager/commands/commands_test.h]: @LuxoftAKutsan May be, but some times, result can depends on `ApplicationManagerSetting` data. So if something in test case went wrong because of unknown call, it will be easier to find a problem.

File: src/components/application_manager/test/commands/command_impl_test.cc (hunk 4 above)
Comment (LevchenkoS) [inline src/components/application_manager/test/commands/command_impl_test.cc]: @AByzhynar Compiler optimization?

File: src/components/application_manager/test/include/application_manager/commands/commands_test.h (hunk 8 above)
Comment (LuxoftAKutsan) [inline src/components/application_manager/test/include/application_manager/commands/commands_test.h]: @LevchenkoS  Settings class is data class, our logic cant strictly depend on calling any methods from settings class. If you find case where you want to test that some class will call any methods in settings class, you probably do something wrong.

File: src/components/application_manager/test/include/application_manager/commands/command_request_test.h (hunk 6 above)
Comment (AByzhynar) [inline src/components/application_manager/test/include/application_manager/commands/command_request_test.h]: @LevchenkoS Yes. `mock_event_dispatcher_` will be even better and clear. But it's not about smbd's opinion. It's about reading of your code by other people. Think about next generations)

File: src/components/application_manager/test/commands/command_impl_test.cc [hunk 10]
```diff
@@ -0,0 +1,285 @@
...
+using ::application_manager::ApplicationManager;
+using ::application_manager::commands::MessageSharedPtr;
+using ::application_manager::ApplicationSharedPtr;
+using ::test::components::application_manager_test::MockApplication;
+
+typedef SharedPtr<MockApplication> MockAppPtr;
+
+class CommandImplTest : public CommandsTest<CommandsTestMocks::kIsNice> {
+ public:
+  class UnwrappedCommandImpl : CommandImpl {
```
Comment (pestOO) [inline src/components/application_manager/test/commands/command_impl_test.cc]: What the reason to wrapping `CommandImpl` in case we are testing it?

File: src/components/application_manager/test/commands/command_impl_test.cc [hunk 11]
```diff
@@ -0,0 +1,285 @@
...
+    UnwrappedCommandImpl(const MessageSharedPtr& message,
+                         ApplicationManager& application_manager)
+        : CommandImpl(message, application_manager) {}
+
+   private:
+    DISALLOW_COPY_AND_ASSIGN(UnwrappedCommandImpl);
+  };
+
+  // Create `SmartObject` which handle array of `SmartObjects`
+  static void CreateArray(MessageSharedPtr& msg,
```
Comment (pestOO) [inline src/components/application_manager/test/commands/command_impl_test.cc]: Look's like `msg` is output parameter.
In this case better to return `MessageSharedPtr` as a result of this method.

Comment (pestOO) [inline src/components/application_manager/test/commands/command_impl_test.cc]: Also, in this case name could be like `CreateArrayMessage`

File: src/components/application_manager/test/commands/command_impl_test.cc (hunk 10 above)
Comment (LevchenkoS) [inline src/components/application_manager/test/commands/command_impl_test.cc]: @pestOO, the reason is to test its protected methods `ReplaceMobileByHMIAppId` and `ReplaceHMIByMobileAppId`

File: src/components/application_manager/test/commands/command_impl_test.cc [hunk 12]
```diff
@@ -0,0 +1,285 @@
...
+                          const size_t msg_count = kDefaultMsgCount_) {
+    if (!msg) {
+      msg = CreateMessage(smart_objects::SmartType_Array);
+    }
+    ::smart_objects::SmartArray* array = msg->asArray();
+    SmartObject obj;
+    for (size_t i = 0u; i < msg_count; ++i) {
+      obj[strings::app_id] = i;
```
Comment (pestOO) [inline src/components/application_manager/test/commands/command_impl_test.cc]: I think it's better to reduce scope of this SO.
In this implementation seems _like_ we append some _tricky_ initialized object. (Yes, I realized that `app_id' is const expretion)

Comment (LevchenkoS) [inline src/components/application_manager/test/commands/command_impl_test.cc]: @pestOO Fixed in aa945bb.

File: src/components/application_manager/test/commands/command_impl_test.cc (hunk 11 above)
Comment (LevchenkoS) [inline src/components/application_manager/test/commands/command_impl_test.cc]: @pestOO Fixed in aa945bb.

File: src/components/application_manager/test/include/application_manager/commands/commands_test.h (hunk 8 above)
Comment (LevchenkoS) [inline src/components/application_manager/test/include/application_manager/commands/commands_test.h]: @LuxoftAKutsan Fixed in aa945bb.

File: src/components/application_manager/test/include/application_manager/commands/command_request_test.h (hunk 6 above)
Comment (LevchenkoS) [inline src/components/application_manager/test/include/application_manager/commands/command_request_test.h]: @AByzhynar Fixed in aa945bb.

File: src/components/application_manager/test/commands/command_impl_test.cc (hunk 4 above)
Comment (LevchenkoS) [inline src/components/application_manager/test/commands/command_impl_test.cc]: @AByzhynar Fixed in aa945bb.

File: src/components/application_manager/test/commands/command_request_impl_test.cc (hunk 2 above)
Comment (LevchenkoS) [inline src/components/application_manager/test/commands/command_request_impl_test.cc]: @AByzhynar Fixed in aa945bb.

**Comment by AByzhynar** [comment]
@LevchenkoS **Wrong branch name**
